# **Neural Collaborative Filtering for Drug prediction**

 

Machine learning can be used in all aspects of human life,

New Drug discovery can be very expensive and can take up to a year, Hence it is very beneficial to find the drug that can be used to treat the specific medical condition from the existing drugs.

This idea is similar to the idea of  recommendation systems where the item is recommended to the user based on certain factors 

Similarly, we can recommend the drug for the treatment of certain medical conditions based on certain factors.

To do so, here we are using the Nural collaborative filtering approach to predict the drugs for a specific medical condition.
We will be using the Multilayer perceptron model as proposed in a research  paper[1]

We have done certain modifications to match the current needs. 

Now let’s start





## **Importing all necessory liberies** 


 

In [ ]:
import sys 
import multiprocessing
from time import time



import numpy as np
import pandas as pd
import random
import math
import argparse
import heapq
import scipy.sparse as sp



import theano
import theano.tensor as T
import tensorflow as tf
import keras

from keras import layers
from keras.models import Sequential,Model
from keras import backend as K
from keras import initializers
from keras.regularizers import l1, l2 
from keras.layers import Dense, Lambda, Activation
from keras.layers import Embedding, Input, Dense, merge ,Reshape, Flatten, Dropout
from keras.optimizers import Adagrad, Adam, SGD, RMSprop





### **Dataset**

we are using drug_com data set which you can download from Kaggle [Here]( https://www.kaggle.com/jessicali9530/kuc-hackathon-winter-2018)


This dataset includes the different drugs and their effectiveness or uses in different medical conditions.

Now let’s analyze the dataset to get an understanding of the data.


In [ ]:
import pandas as pd
df = pd.read_csv('/content/drive/My Drive/76158_171475_compressed_drugsComTrain_raw.csv.zip').drop(['uniqueID','date'],axis = 1)
df

drugName  ... usefulCount
0                      Valsartan  ...          27
1                     Guanfacine  ...         192
2                         Lybrel  ...          17
3                     Ortho Evra  ...          10
4       Buprenorphine / naloxone  ...          37
...                          ...  ...         ...
161292                   Campral  ...         125
161293            Metoclopramide  ...          34
161294                   Orencia  ...          35
161295        Thyroid desiccated  ...          79
161296              Lubiprostone  ...         116

[161297 rows x 5 columns]

In [ ]:
print(len(df['condition'].unique()))
print(len(df['drugName'].unique()))

885
3436


In [ ]:
Medical_Conditions = df['condition'].unique()
Drugs = df['drugName'].unique()

.


Our dataset contains the drug name and respective medical condition, with all other information like a review , rating, useful count ...
 
Now we want to find the usefulness of a particular drug in the different conditions ( i.e predicting the drug  for certain condition)
 
Here we can see that we have 3436 unique drugs with 885 unique conditions and also 161297 interactions.



### **Mapping the drug names to Unique IDs and Unique IDs to  drug names**

In [ ]:
Medical_Conditions_ID_to_name= {}

for i in range(len(Medical_Conditions)):
  key = i
  value = Medical_Conditions[i]
  Medical_Conditions_ID_to_name[key] = value

Drugs_ID_to_name = {}


for i in range(len(Drugs)):
  key = i
  value = Drugs[i]
  Drugs_ID_to_name[key] = value

Medical_Conditions_ID_to_NAME = pd.DataFrame(list(Medical_Conditions_ID_to_name.items()))
Drugs_ID_to_NAME = pd.DataFrame(list(Drugs_ID_to_name.items()))

Medical_Conditions_ID_to_NAME.to_csv('Medical_Conditions_ID_to_NAME.csv')
Drugs_ID_to_NAME.to_csv('Drugs_ID_to_NAME.csv')

In [ ]:
Medical_Conditions_name_to_ID =  dict([(value, key) for key, value in Medical_Conditions_ID_to_name.items()])
Drugs_name_to_ID =  dict([(value, key) for key, value in Drugs_ID_to_name.items()])

Considering the useful columns


(Here we are using the usefulcount as the interaction between drug and specific medical condition )


In [ ]:
df = df[['condition','drugName','usefulCount']].copy()
df

condition                  drugName  usefulCount
0       Left Ventricular Dysfunction                 Valsartan           27
1                               ADHD                Guanfacine          192
2                      Birth Control                    Lybrel           17
3                      Birth Control                Ortho Evra           10
4                  Opiate Dependence  Buprenorphine / naloxone           37
...                              ...                       ...          ...
161292            Alcohol Dependence                   Campral          125
161293               Nausea/Vomiting            Metoclopramide           34
161294          Rheumatoid Arthritis                   Orencia           35
161295           Underactive Thyroid        Thyroid desiccated           79
161296         Constipation, Chronic              Lubiprostone          116

[161297 rows x 3 columns]

Mapping

In [ ]:
for i in range(len(df['drugName'])):
  df['drugName'][i] = Drugs_name_to_ID[df['drugName'][i]]

for i in range(len(df['condition'])):
  df['condition'][i] = Medical_Conditions_name_to_ID[df['condition'][i]]

df.sort_values("condition", axis = 0, ascending = True, 
                 inplace = True)

df

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """


condition drugName  usefulCount
0              0        0           27
45609          0      147           16
2044           0      667           31
26551          0      887           43
113009         0      887           50
...          ...      ...          ...
159071       880     1970            1
159319       881       93           11
160791       882      420           62
160921       883      470           92
161018       884     2874           25

[161297 rows x 3 columns]

As this is a binary classification problem we are not considering the properties of drugs ( can be done in the next version )

We are going to build the sparse matrix for representing interactions between drugs and medical conditions 

For testing, we have taken 200 users with their positive, negative interactions (Please have a look at "log loss with negative sampling" part in the referred research paper[1]



In [ ]:
def get_data():
  '''
  return :  train and test data
  train-> sparce matrix
  test-> matrix (list of lists)
  '''

  train = sp.dok_matrix((885, 3436), dtype=np.float32)
 
  for i in  range(len(df['condition'])):
    ls = list(df.iloc[i])
    train[ls[0],ls[1]] = 1.0


  test = []

  for j in  range(200):
    i = random.randint(0, 161296)
    ls = list(df.iloc[i])
    test.append([ls[0],ls[1]])

  return train , test 
  


### **Evaluation Process**

To evaluate the performance of
drug recommendation, we adopted the leave-one-out evaluation.
as described in the referred research paper[1]
 
HR intuitively measures whether
the test item is present on the top-10 list, and the NDCG
accounts for the position of the hit by assigning higher scores
to hits at top ranks. We will calculate both metrics for each
test user and asssine the average score.


In [ ]:
def getHitRatio(ranklist, gtItem):
    for item in ranklist:
        if item == gtItem:
            return 1
    return 0

def getNDCG(ranklist, gtItem):
    for i in range(len(ranklist)):
        item = ranklist[i]
        if item == gtItem:
            return math.log(2) / math.log(i+2)
    return 0
    

In [ ]:
def evaluate(model,K):

  HR, NDCG = [],[]
  train,test = get_data()


  for i in range(len(test)):                                              

    rating = test[i]                                                     
    u = rating[0]                                                       

    # taking 99 randome untested conditions by that drug
    count = 0
    drugs = [] 
    while(count != 99):
      j = random.randint(0, 3435)     
      if (u,j) in train.keys():
        continue
      drugs.append(j)
      count+=1
                                 
                                                  
                                                         
    gtdrug = rating[1]                                                     
    drugs.append(gtdrug)                                                   
    
    
    
    # Get prediction scores
    map_drug_score = {}
    medical_conditions = np.full(len(drugs), u, dtype = 'int32')
    predictions = model.predict([medical_conditions, np.array(drugs)], 
                                 batch_size=64, verbose=0)               
    
    
    for i in range(len(drugs)):                                            
        drug = drugs[i]
        map_drug_score[drug] = predictions[i]

    drugs.pop()                                                             
    
     
    ranklist = heapq.nlargest(K, map_drug_score, key=map_drug_score.get)   
    hr = getHitRatio(ranklist, gtdrug)                                       
    ndcg = getNDCG(ranklist, gtdrug)

    HR.append(hr)
    NDCG.append(ndcg) 

  return (HR, NDCG)






### **Prepareing model**


**Multi-Layer Presepteron** :( as described in the paper[1])
 
Since NCF adopts two pathways to model drugs and conditions,
it is intuitive to combine the features of two pathways by
concatenating them. This design has been widely adopted
in multimodal deep learning works. However, simply
a vector concatenation does not account for any interactions
between drug and condition latent features, which is insufficient
for modeling the collaborative filtering effect. To address
this issue, we will add hidden layers on the concatenated vector, using a standard MLP to learn the interaction
between drug and condition latent features. In this sense, we can
endow the model a large level of **flexibility and non-linearity.**


In [ ]:
def get_model(num_medical_conditions, num_drugs, layers = [16,8], reg_layers=[0,0]):

    assert len(layers) == len(reg_layers)

    num_layer = len(layers)                                                                      #Number of layers in the MLP
    
    medical_condition_input = Input(shape=(1,), dtype='int32', name = 'user_input')
    drug_input = Input(shape=(1,), dtype='int32', name = 'item_input')


    MLP_Embedding_Medical_Conditions = Embedding(input_dim = num_medical_conditions, output_dim = int(layers[0]/2), 
                                   name = 'medical_condition_embedding',
                                    W_regularizer = l2(reg_layers[0]), input_length=1)
    MLP_Embedding_Drugs = Embedding(input_dim = num_drugs, output_dim = int(layers[0]/2), 
                                   name = 'drug_embedding',
                                    W_regularizer = l2(reg_layers[0]), input_length=1)   
    

   
    medical_condition_latent = Flatten()(MLP_Embedding_Medical_Conditions(medical_condition_input))                                      # flattening embedding for user 
    drug_latent = Flatten()(MLP_Embedding_Drugs(drug_input))                                      # flattening embedding for items
     

    vector = keras.layers.concatenate([medical_condition_latent,drug_latent])                                 # forming the 0th layer of NN by concatinating the user and items flatten layer 
    
    # MLP layers
    for idx in range(1, num_layer):
        layer = Dense(layers[idx], W_regularizer= l2(reg_layers[idx]), activation='relu', name = 'layer%d' %idx)
        vector = layer(vector)
        #layer1 = Dropout(0.25)
        #vector = layer1(vector)
        
    
    
        
    # Final prediction layer
    prediction = Dense(1, activation='sigmoid', init='lecun_uniform', name = 'prediction')(vector)
    
    model = Model(input=[medical_condition_input, drug_input], 
                  output=prediction)
    
    return model



In [ ]:

def get_train_instances(train, num_negatives):
  
    medical_condition_input, drug_input, labels = [],[],[]
    num_medical_conditions = train.shape[0]

    for (u, i) in train.keys():

        # positive instance
        medical_condition_input.append(u)
        drug_input.append(i)
        labels.append(1)


        # negative instances
        for t in range(num_negatives):
            j = np.random.randint(num_drugs)
            
            while ( (u,j) in train.keys() ) :
                j = np.random.randint(num_drugs)
            medical_condition_input.append(u)
            drug_input.append(j)
            labels.append(0)
        
        
    return medical_condition_input, drug_input, labels

### **Training testing and evaluating  :**

The Neural network architecture that will have 6 hidden layers with one input layer(formed by concatenating drugs and conditions embeddings) and one output layer


the optimizer, learning rate, batch size, epochs ... are decided after doing few experiments, and optimum is chosen for this work.


The process includes :
1. create a model
2. train model
3. test/ evaluate model
4. calculate HR, NDCG
5. check for the best HR, NDCG and save the model
6. Repeat  the steps 3, 4 and 5 for "epochs" times

This will save the best model 


In [ ]:

if __name__ == '__main__':

    path = '/content'
    layers = [256,128,64,32,16,8]
    reg_layers = [0,0,0,0,0,0]
    num_negatives =  6
    learner =  'adam'
    learning_rate = 0.001
    batch_size = 256
    epochs = 10
    verbose = 1
    
    topK = 5
    model_out_file = 'Pretrain_new.h5'  
    
    # Loading data
    t1 = time()
    train,test = get_data()

    num_medical_conditions, num_drugs = train.shape
    
    # Build model
    model = get_model(num_medical_conditions, num_drugs, layers, reg_layers)


    #compile model
    model.compile(optimizer=Adam(lr=learning_rate), loss='binary_crossentropy',metrics=['accuracy'])


    
    # Check Init performance
    t1 = time()
    (hr, ndcg) = evaluate(model,topK)
    HR, NDCG = np.array(hr).mean(), np.array(ndcg).mean()
    print('Init: HR = %.4f, NDCG = %.4f [%.1f]' %(HR, NDCG, time()-t1))
    


    # Train model
    best_hr, best_ndcg, best_iter = HR, NDCG, -1


    for epoch in range(epochs):
        t1 = time()

        # Generate training instances
        medical_condition_input, drug_input, labels = get_train_instances(train, num_negatives)
    

        # Training        
        hist = model.fit([np.array(medical_condition_input), np.array(drug_input)],
                         np.array(labels),batch_size=batch_size, nb_epoch=20, verbose=0, shuffle=True)
        '''plt.plot(history.history["loss"])
        #plt.plot(history.history["val_loss"])
        plt.title("model loss")
        plt.ylabel("loss")
        plt.xlabel("epoch")
        plt.legend(["train", "test"], loc="upper left")
        plt.show()'''
        
        t2 = time()

        # Evaluation
        if epoch %verbose == 0:

            (hr, ndcg) = evaluate(model,topK)
            HR, NDCG, loss = np.array(hr).mean(), np.array(ndcg).mean(), hist.history['loss'][0]
            print('Iteration %d [%.1f s]: HR = %.4f, NDCG = %.4f, loss = %.4f [%.1f s]' 
                  % (epoch,  t2-t1, HR, NDCG, loss, time()-t2))
            
            if HR >= best_hr and NDCG >= best_ndcg:

                best_hr, best_ndcg, best_iter = HR, NDCG, epoch
                model.save(model_out_file)



    print("End. Best Iteration %d:  HR = %.4f, NDCG = %.4f. " %(best_iter, best_hr, best_ndcg))
    print("The best MLP model is saved to %s" %(model_out_file))

/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:13: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=885, output_dim=128, name="medical_condition_embedding", input_length=1, embeddings_regularizer=<keras.reg...)`
  del sys.path[0]
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:16: UserWarning: Update your `Embedding` call to the Keras 2 API: `Embedding(input_dim=3436, output_dim=128, name="drug_embedding", input_length=1, embeddings_regularizer=<keras.reg...)`
  app.launch_new_instance()
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:28: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(128, activation="relu", name="layer1", kernel_regularizer=<keras.reg...)`
/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:28: UserWarning: Update your `Dense` call to the Keras 2 API: `Dense(64, activation="relu", name="layer2", kernel_regularizer=<keras.reg...)`
/usr/local/lib/python3.6/dist-packages/

Init: HR = 0.0300, NDCG = 0.0204 [40.9]


/usr/local/lib/python3.6/dist-packages/ipykernel_launcher.py:53: UserWarning: The `nb_epoch` argument in `fit` has been renamed `epochs`.
/usr/local/lib/python3.6/dist-packages/tensorflow/python/framework/indexed_slices.py:434: UserWarning: Converting sparse IndexedSlices to a dense Tensor of unknown shape. This may consume a large amount of memory.
  "Converting sparse IndexedSlices to a dense Tensor of unknown shape. "


Iteration 0 [33.2 s]: HR = 1.0000, NDCG = 0.8826, loss = 0.4345 [40.3 s]
Iteration 1 [32.0 s]: HR = 1.0000, NDCG = 0.9490, loss = 0.0976 [40.4 s]
Iteration 2 [31.8 s]: HR = 1.0000, NDCG = 0.9717, loss = 0.0586 [40.2 s]
Iteration 3 [31.7 s]: HR = 1.0000, NDCG = 0.9809, loss = 0.0471 [40.1 s]
Iteration 4 [32.5 s]: HR = 1.0000, NDCG = 0.9926, loss = 0.0337 [40.7 s]
Iteration 5 [31.7 s]: HR = 1.0000, NDCG = 0.9852, loss = 0.0338 [40.3 s]
Iteration 6 [31.8 s]: HR = 1.0000, NDCG = 0.9898, loss = 0.0284 [40.2 s]
Iteration 7 [31.8 s]: HR = 1.0000, NDCG = 0.9871, loss = 0.0283 [40.3 s]
Iteration 8 [32.6 s]: HR = 1.0000, NDCG = 0.9815, loss = 0.0228 [39.8 s]
Iteration 9 [31.9 s]: HR = 1.0000, NDCG = 0.9852, loss = 0.0192 [39.8 s]
End. Best Iteration 4:  HR = 1.0000, NDCG = 0.9926. 
The best MLP model is saved to Pretrain_new.h5


In [ ]:
Best_model = tf.keras.models.load_model(
    '/content/Pretrain_new.h5', compile = True
)

Now let’s try to find out the possible drugs for Rheumatoid Arthritis( Rheumatoid arthritis (RA) is an autoimmune disease that can cause joint pain and damage throughout your body)

In [ ]:

drugs = [i for i in range(3436)]

medical_conditions = np.full(len(drugs),18, dtype = 'int32')

predictions = Best_model.predict([medical_conditions, np.array(drugs)], 
                                 batch_size=100, verbose=0)

map_drug_score ={}

for i in range(len(drugs)):                                           # creating the{ item : chance } dict 
        drug = drugs[i]
        map_drug_score[drug] = predictions[i]

ranklist = heapq.nlargest(5, map_drug_score, key=map_drug_score.get)



print( str(Medical_Conditions_ID_to_name[18]) +" can be treated by : " )

print("\n")
for i in ranklist:
  print("\n\t"+str(Drugs_ID_to_name[i]))


Rheumatoid Arthritis can be cured by : 



	Etanercept

	Acetaminophen / hydrocodone

	Hydroxychloroquine

	Naproxen

	Aleve


The model has predicted a few correct drugs like:
* **Hydroxychloroquine**

and also suggested possible drugs for the treatment of Rheumatoid arthritis
* Etanercept

* Acetaminophen / hydrocodone

* Naproxen

* Aleve


Although the working of these drugs can only be confirmed by the **scientists working in these fields**.
 




This was the basic model for drug prediction 

Modification can be done to lear some more complex relations 

  Few modifications : 
  

*   We can consider the featchers of the drug ( like protine structure , activity in different environment ... )
*   large data set can also help to increse the radius of the possibilities and can lead to better predictions 
* using more deep NN structure 
* fine tuning the hyperparameters 
* ...

In the second part we will try to create such model and find the possibilites for COVID-19 ...  :) 






## **References **: 

[1] Xiangnan He, Lizi Liao, Hanwang Zhang, Liqiang Nie, Xia Hu and Tat-Seng Chua (2017). Neural Collaborative Filtering. In Proceedings of WWW '17, Perth, Australia, April 03-07, 2017.


